### PR TITLE
Make jumps to timers faster

### PR DESCRIPTION
**Pull request by akopich**
With the new generalised way of treating timers of all types
an efficient implementation of min/max/first/last timer
search has became much more difficult. In that regard we
pre-calculate min/max/first/last timers for each scope_id
upon capture finalisation.

Tests: Manual
Bug: http://b/229731732

### CODE REVIEW

File: src/ClientData/include/ClientData/CaptureData.h [hunk 1]
```diff
@@ -244,6 +244,32 @@ class CaptureData {
   [[nodiscard]] const std::vector<uint64_t>* GetSortedTimerDurationsForScopeId(
       uint64_t scope_id) const;
 
+  // TODO(b/205037572) This is a hack allowing for fast jumps to timers. The class should be removed
+  // after the way we store the `timer_info`s is refactored.
+  class FirstLastMinMaxTimers {
```
Comment (dpallotti) [inline src/ClientData/include/ClientData/CaptureData.h:249]: I think this should just be in its own file.

File: src/ClientData/CaptureData.cpp [hunk 2]
```diff
@@ -222,4 +237,19 @@ void CaptureData::UpdateTimerDurations() {
   return result;
 }
 
+void CaptureData::FirstLastMinMaxTimers::Update(const TimerInfo* new_timer) {
+  if (first == nullptr || first->end() > new_timer->end()) {
```
Comment (dpallotti) [inline src/ClientData/CaptureData.cpp:241]: Isn't it guaranteed that `first`, `last`, ... are not `nullptr`?

File: src/ClientData/include/ClientData/CaptureData.h [hunk 3]
```diff
@@ -244,6 +244,32 @@ class CaptureData {
...
+          last(first_timer_observed),
+          min(first_timer_observed),
+          max(first_timer_observed) {}
+
+    void Update(const TimerInfo* new_timer);
+
+    const TimerInfo* first;
+    const TimerInfo* last;
+    const TimerInfo* min;
+    const TimerInfo* max;
```
Comment (dpallotti) [inline src/ClientData/include/ClientData/CaptureData.h:262]: These should be private, only `Update` should be able to modify them.

File: src/OrbitGl/App.cpp [hunk 4]
```diff
@@ -2808,32 +2808,29 @@ bool OrbitApp::HasFrameTrackInCaptureData(uint64_t instrumented_function_id) con
...
   switch (selection_mode) {
     case JumpToTimerMode::kFirst: {
-      const auto* first_timer = GetMutableTimeGraph()->FindNextScopeTimer(
-          scope_id, std::numeric_limits<uint64_t>::lowest());
-      if (first_timer != nullptr) GetMutableTimeGraph()->SelectAndZoom(first_timer);
+      timer_to_zoom = first_last_min_max_timers->first;
       break;
     }
     case JumpToTimerMode::kLast: {
-      const auto* last_timer = GetMutableTimeGraph()->FindPreviousScopeTimer(
```
[the remark this answers is not in the document]
Comment (akopich) [reply src/OrbitGl/App.cpp:2819]: Outstanding point. This hack is incomplete then. Closing the PR. @dpallotti , sorry you had to review this.

**Comment by karupayun** [comment]
> Outstanding point. This hack is incomplete then. Closing the PR. @dpallotti , sorry you had to review this.

In this PR, you are making the jump to first/last/max/min work in O(1), but paying a cost in post-processing. I think we might be interested in this idea in the future where we might have 2 versions of the capture, the initial one and the optimized one. In this moment, there is a general consensus that we don't want to pay any more extra cost in post-processing.

**Pull request closed without merging**